src/controller/python/chip/storage/delegate.py

```diff
@@ -0,0 +1,131 @@
+            res = cur.fetchone()
+            return base64.standard_b64decode(res[0]) if res is not None else None
+
+    def Delete(self, domain:str, field:str, key:str)->None:
+        with self.lock:
+            cur = self.con.cursor()
+            cur.execute("DELETE FROM storage WHERE domain = ? AND field = ? AND key = ?", (domain, field, key))
+            self.con.commit()
+
+_controllerStorage = None
+
+@_ControllerGetKeyValueFunct
+def ControllerGetKeyValueFunct(key:c_char_p, keyLen:int, valPtr:c_void_p, sizePtr:ctypes.POINTER(c_uint16)):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    data = _controllerStorage.Get(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey)
+    sizeRead = ctypes.pointer(c_uint16(len(data))) if data is not None else ctypes.pointer(c_uint16(0xffff))
+    if data is not None and valPtr != 0:
+        copySize = min(len(data), sizePtr.contents.value)
+        ctypes.memmove(valPtr, ctypes.c_char_p(data), copySize)
+    ctypes.memmove(sizePtr, sizeRead, ctypes.sizeof(c_uint16))
+
+@_ControllerSetKeyValueFunct
+def ControllerSetKeyValueFunct(key:c_char_p, keyLen: int, valPtr:c_void_p, size:int):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    data = ctypes.string_at(valPtr, size)
+    _controllerStorage.Set(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey, data)
+
+@_ControllerDeleteKeyValueFunct
+def ControllerDeleteKeyValueFunct(key:c_char_p, keyLen: int):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    _controllerStorage.Delete(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey)
+
+def _InitControllerStorageInternal(storageClass, options: StorageOptions):
+    global _controllerStorage
+    handle = chip.native.GetLibraryHandle()
+    _controllerStorage = storageClass(options)
+    if not handle.pychip_PythonPersistentStorageDelegate_SetCallbacks.argtypes:
+        setter = chip.native.NativeLibraryHandleMethodArguments(handle)
+        setter.Set("pychip_PythonPersistentStorageDelegate_SetCallbacks", None, [_ControllerGetKeyValueFunct, _ControllerSetKeyValueFunct, _ControllerDeleteKeyValueFunct])
+
+        handle.pychip_PythonPersistentStorageDelegate_SetCallbacks(ControllerGetKeyValueFunct, ControllerSetKeyValueFunct, ControllerDeleteKeyValueFunct)
+
+def InitControllerStorage(storageUrl: str):
+    storageMap = {
+        "sqlite3": SqlStorage
+    }
+    storage = urlparse.urlparse(storageUrl)
+    storageClass = storageMap.get(storage.scheme, None)
+    if not storageClass:
+        raise Exception("Unknown storage: {}".format(storage.scheme))
+    options = StorageOptions(path=storage.path, params=urlparse.parse_qs(storage.query))
+    _InitControllerStorageInternal(storageClass, options)
```

Comment:
Why is there an internal and external init?

Pattern in other chip.native files is to have a 'get library handle' that returns an initialized handle. Please follow that for consistency (i.e. do not do both handle initialization and storage class initialization in the same init function)